platforms/renderer-generic-egl: Support CPUAddressableDisplayProvider

RAOF opened the pull request:
Fixes: #3092

RAOF commented:
(Don't yet know if it actually fixes that; putting up a PR so that I can get a snap to test :wink: )

RAOF commented:
Hm, no, doesn't fix it, because `eglGeyDisplay(EGL_DEFAULT_DISPLAY)` only tries X11 (on mesa).

Time to pull out [`EGL_MESA_platform_surfaceless`](https://registry.khronos.org/EGL/extensions/MESA/EGL_MESA_platform_surfaceless.txt)

mattkae commented:
> Hm, no, doesn't fix it, because `eglGeyDisplay(EGL_DEFAULT_DISPLAY)` only tries X11 (on mesa).
> 
> Time to pull out [`EGL_MESA_platform_surfaceless`](https://registry.khronos.org/EGL/extensions/MESA/EGL_MESA_platform_surfaceless.txt)

The following makes us not fail _right away_:
```c++
auto dpy = eglGetPlatformDisplay(EGL_PLATFORM_SURFACELESS_MESA, EGL_DEFAULT_DISPLAY, NULL);
```

But you'll start getting some funny errors like:
![image](https://github.com/MirServer/mir/assets/25062299/8e280fdd-821a-4a3b-b9c6-d1a4a80a1e7c)

Maybe you know something I don't know there :) [This post](https://www.reddit.com/r/opengl/comments/14xovxn/egl_mesa_platform_surfaceless_question/) had some good info

RAOF commented:
There we go, that should be good to go. I just:
* Fixed up the `eglGetPlatformDisplay` usage - we need to go through the extension, not through the EGL 1.5 entrypoint (because at least one driver we want to support does not provide EGL 1.5 entrypoints)
* Added a bunch of extension checks, to validate that what we're doing is right.

AlanGriffiths commented on include/platform/mir/graphics/egl_extensions.h:
```diff
@@ -214,6 +214,9 @@ namespace mir
 {
 namespace graphics
 {
+auto has_egl_client_extension(char const* name) -> bool;
+auto has_egl_extension(EGLDisplay dpy, char const* name) -> bool;
```
Nit: I think `extension` would be bit more specific than `name`

AlanGriffiths commented on src/platform/graphics/egl_extensions.cpp:
```diff
@@ -15,12 +15,32 @@
+
+auto mg::has_egl_extension(EGLDisplay dpy, char const* name) -> bool
+{
+    auto extensions = eglQueryString(dpy, EGL_EXTENSIONS);
+    auto found_substring = strstr(extensions, name);
```
could be const

AlanGriffiths commented on src/platform/graphics/egl_extensions.cpp:
```diff
@@ -15,12 +15,32 @@
+{
+    auto extensions = eglQueryString(dpy, EGL_EXTENSIONS);
+    auto found_substring = strstr(extensions, name);
+    if (found_substring)
+    {
+        // Check that we haven't found a prefix of our extension name
+        auto end_of_match = found_substring + strlen(name);
+        // It's a match if it terminates with the end of the extension string, or with a space
+        return *end_of_match == '\0' || *end_of_match == ' ';
+    }
+    return false;
```
I think this is wrong in theory: There Could be a later entry that matches.

AlanGriffiths reviewed (changes requested):
Looks like a good approach, but there are a few bits to tidy up

Saviq commented:
So this isn't yet stable:

```shell
$ MIR_SERVER_PLATFORM_DISPLAY_LIBS=mir:virtual MIR_SERVER_PLATFORM_RENDERING_LIBS=mir:egl-generic MIR_SERVER_VIRTUAL_OUTPUT=1280x1024 ./bin/mir_performance_tests
```

I'm getting this on the CompositorPerformance test:
```plain
[ RUN      ] CompositorPerformance.regression_test_1563287
Saving server logs to: /tmp/CompositorPerformance_regression_test_1563287_server.log
/home/michal/dev/MirServer/mir/tests/performance-tests/test_compositor.cpp:102: Failure
Expected: (compositor_fps) >= (0), actual: -1 vs 0
/home/michal/dev/MirServer/mir/tests/performance-tests/test_compositor.cpp:103: Failure
Expected: (compositor_render_time) > (0), actual: -1 vs 0
[  FAILED  ] CompositorPerformance.regression_test_1563287 (14611 ms)
```

And this, once…
```
[ RUN      ] HostedGLMark2Wayland.windowed                                                                                         
Saving host output to: /tmp/HostedGLMark2Wayland_windowed_host.log
Saving server logs to: /tmp/HostedGLMark2Wayland_windowed_server.log
Mir fatal error: Critical error in Wayland platform: /home/michal/dev/MirServer/mir/src/platforms/wayland/display.cpp(250): Throw in function void mir::graphics::wayland::Display::run()
Dynamic exception type: boost::wrapexcept<std::system_error>
std::exception::what: Failed to read Wayland events: Broken pipe
```

Neither are a blocker, FWIW.

AlanGriffiths commented:
> Expected: (compositor_render_time) > (0), actual: -1 vs 0

This is because we don't get any output from `-compositor-report=log`: #3101

Saviq reviewed (approved):
> Neither are a blocker, FWIW.

I've confirmed both on main, so good with me!

AlanGriffiths commented:
> > Expected: (compositor_render_time) > (0), actual: -1 vs 0
> 
> This is because we don't get any output from `-compositor-report=log`: #3101

But now I do:

```
$ grep compositor: /tmp/CompositorPerformance_regression_test_1563287_server.log
[2023-10-26 14:00:36.827435] <information> compositor: Started
[2023-10-26 14:00:48.289107] <information> compositor: Stopped
```

Was I mad then? Or am I mad now?

sergio-costas commented:
It doesn't crash on Qemu, but still doesn't work correctly:

![imagen](https://github.com/MirServer/mir/assets/105285003/21bbc5b9-b50f-46a3-9491-770a976c21e0)

sergio-costas commented:
(launched with `miriway-shell --display-autoscale=1200 --shell-component="swaybg -c #2C001E"` )
